### PR TITLE
Fully specialize Bitboard attacks_bb and attacks_from. This remove un…

### PR DESCRIPTION
**Pull request by OuaisBla**
Commit | Specialization_attacks_bb (Fully specialize Bitboard attacks_bb and attacks_from. This remove unused attacks_bb and attacks_bb template instantiation.)
-- | --
Info | Fully specialize Bitboard attacks_bb and attacks_from. This remove unused attacks_bb and attacks_bb template instantiation.
Submitter | OuaisBla
TC | 10+0.1
SPRT | elo0: -1.50  alpha: 0.05  elo1: 0.50  beta: 0.05 (logistic)
LLR | 2.95 [-2.94,2.94] (accepted)
Elo | 2.26 [-0.68,5.16] (95%)
LOS | 93.5%
Games | 19536 [w:19.5%, l:18.8%, d:61.7%]
Link | /tests/view/5e5ddf13e42a5c3b3ca2e06f

https://tests.stockfishchess.org/tests/view/5e5ddf13e42a5c3b3ca2e06f

To explain a bit more, the attacks_bb and attacks_bb function are instianted by the pos.attack_from and pos.attack_from method respectively. The base code rely on conditional operation to exclude the calls leaving the choice to the compiler to exclude it or not in its binary.

The specialized template versions of attack_bb and pos. attack_from don't create the above functions at all and make the life easier for the compiler to make possibly better inlining decision.

Positive fishtest result and binary size change on my PC.

Base binary size: 558 Ko (572 095 octets)
New binary size: 557 Ko (570 553 octets)

Build using Cygwin and make ARCH=x86-64 COMP=gcc all

### CODE REVIEW

**Comment by protonspring** [comment]
i don't think i would add this unless it demonstrably faster than master. Usually adding lots of code like this requires to pass STC -0.5 / 1.5.

**Comment by OuaisBla** [comment]
Please beware that the notion of adding code when it comes to template instantiation vs explicit template specialization can be confusing. With template, you simply rely on the compiler to generate it for you. In fact, there is less code generated by specializing more in this case, hence the diff in binary size. So, with this patch, I basically remove template instiantied code and binary generated code, while adding few written line of code. That has some merits in my opinion, plus I can understand the new code more easily because it is more readable for me. 

The simplification [-1.5, 0.5] converged rapidely and LOS was at 86% or so. from a maintainer perspective, does the current test is enough? I don't mind a STC -0.5 / 1.5 but doesn't want to spend fishtest ressources if not necessary. This patch is obvious in my opinion.

**Comment by OuaisBla** [comment]
Results for 50 tests for each version:

  | Base | Test | Diff
-- | -- | -- | --
Mean | 1507545 | 1502694 | 4851
StDev | 21268 | 24151 | 6720

p-value: 0,235
speedup: -0,003

Fishbench of 50 iterations shows a slight speed up of 0.3%.

**Comment by protonspring** [comment]
I also did the same thing to pawn_double_attacks_bb and tested your patch and don't detect any speed differences at all (< 0.1% difference).  This makes sense, because the templated functions in master should make all of the same methods.

I also note that we lost the PAWN exclusion, so you'll have to add a method for PAWNs to notify the user that this method cannot be used for PAWNs.  This is a lot of extra code for presumably no benefit.

On my machine, this patch makes a slightly larger binary: 358744 vs 358632 for master.

**Comment by JeanGauthierOps** [comment]
@protonspring I did another test too on a second machine. The binary are the same in that second case. The difference is the g++ version used. With the older G++ version (6.4.0), the binary size are effectively different, but not with the newest one  (9.2.0). My FishBench results are therefore biaised according to the older G++ compiler.

So, this is still consistent to whether a compiler will be able to inline a given code or not. The new compiler seems able to unroll and wipe the conditional expression while the older one (6.4.0) seems to have more difficulty to do it.

I would personally write code that helps the compiler by making it as explicit as possible, although is still ok to let it work for you and find optimization by itself.

So basically, this patch seems to help generating faster code for older compiler only.. I don't know why Fishtest seems to have detect a LOS of 93%. What it the gcc version deployed to compile the code locally by fishtest tooling? My G++ run by Fishtest dates back 2016. That could explain the high LOS too if all client run similar G++ compiler version. If it were not the positive fishtest results, I would have been suspicious too regarding benefits of that patch.

**Comment by OuaisBla** [comment]
@protonspring I did another test too on a second machine. The binary are the same in that second case. The difference is the g++ version used. With the older G++ version (6.4.0), the binary size are effectively different, but not with the newest one (9.2.0). My FishBench results are therefore biaised according to the older G++ compiler.

So, this is still consistent to whether a compiler will be able to inline a given code or not. The new compiler seems able to unroll and wipe the conditional expression while the older one (6.4.0) seems to have more difficulty to do it.

I would personally write code that helps the compiler by making it as explicit as possible, although is still ok to let it work for you and find optimization by itself.

So basically, this patch seems to help generating faster code for older compiler only.. I don't know why Fishtest seems to have detect a LOS of 93%. What it the gcc version deployed to compile the code locally by fishtest tooling? My G++ run by Fishtest dates back 2016. That could explain the high LOS too if all client run similar G++ compiler version. If it were not the positive fishtest results, I would have been suspicious too regarding benefits of that patch.

**Comment by OuaisBla** [comment]
Confirmed. The G++ provided by FishtestWorker package is:

gcc version 6.4.0 (GCC)

Same as the one I was using on my first machine with Cygwin. Using a newer g++ compiler like 9.2.0. yield better optimization and clear the effect of this patch.

**Comment by OuaisBla** [comment]
Current status:

GCC 6.4.0: Binary size are different and FishBench measure an increase in bench speed for the Patch, confirmed by an ELO gain by FishTest.

GCC 9.2.0: Binary size are equal, due to a better compiler.

I know that @protonspring probably won't like it.. but I would recommand to integrate this patch for this reasons: FishTest run mostly on 6.4.0 and this patch yield better and faster code. So unless we upgrade FishTest, not accepting this patch means that we are wasting Fishtest computer power up to 0.8%.

**Comment by ddugovic** [comment]
How does fully specializing `pawn_attacks_bb` improve performance?

If the goal is to reduce waste on fishtest and we're committed to using GCC 6, are there GCC 6 features we should consider? https://gcc.gnu.org/gcc-6/changes.html#cxx

**Comment by OuaisBla** [comment]
@ddugovic This is related to the inline optimization, which is when the compiler remove the fonction prologue and epilogue of the stack mamagement + the indirection of the function call itself. If the compiler is not sure about possible side effect of inlining a function, it simply won't do it. The "inline" keyword is only a hint, not a prescription nor an obligation.

Compilers have always improve the inlining decision process over time. 

In the actual case, the conditional operator on the PieceType template argument seems to prevent such optimization on the GCC 6.4.0 compiler. But not for the GCC 9.2.0 version which seems definitely better. 

Since those attacks_bb<PieceType> method are spread all around in the code, Better inlining decision by the compiler make better binary and faster code overall. That seems significant to consider doing something at this moment for this. Either include the patch or improve the compiler version or settings.

For GCC, some option could be investigate, like from the https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html#Optimize-Options

max-inline-insns-recursive
max-inline-insns-recursive-auto
Specifies the maximum number of instructions an out-of-line copy of a self-recursive inline function can grow into by performing recursive inlining.

--param max-inline-insns-recursive applies to functions declared inline. For functions not declared inline, recursive inlining happens only when -finline-functions (included in -O3) is enabled; --param max-inline-insns-recursive-auto applies instead.

max-inline-recursive-depth
max-inline-recursive-depth-auto
Specifies the maximum recursion depth used for recursive inlining.

--param max-inline-recursive-depth applies to functions declared inline. For functions not declared inline, recursive inlining happens only when -finline-functions (included in -O3) is enabled; --param max-inline-recursive-depth-auto applies instead.

min-inline-recursive-probability
Recursive inlining is profitable only for function having deep recursion in average and can hurt for function having little recursion depth by increasing the prologue size or complexity of function body to other optimizers.

When profile feedback is available (see -fprofile-generate) the actual recursion depth can be guessed from the probability that function recurses via a given call expression. This parameter limits inlining only to call expressions whose probability exceeds the given threshold (in percents).

**Comment by vondele** [comment]
@OuaisBla I don't think this is worth adding. It is not a simplification from a source code point of view.

A speedup would surprise me, I don't think there is anything there that a compiler can't do. If it passes our usual STC and LTC Elo gainer bounds, I would reconsider.

Concerning the version of gcc used in fishtest, most workers (noob), are on Linux, where it depends what version of gcc the user has installed.
